JapicmpTask.collectArchives crashes on BOM dependency

We have a [PR in Wire](https://github.com/square/wire/pull/2103) switching our dependency on Okio to use a [BOM](https://docs.gradle.org/current/userguide/platforms.html#sub:bom_import).

It looks like the japicmp plugin (`0.3.0`) doesn't like it and crashes with the following:

```
Execution failed for task ':wire-runtime:japicmp:japicmp'.
> java.util.NoSuchElementException (no error message)

* Try:
Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':wire-runtime:japicmp:japicmp'.
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:188)
        .....
Caused by: java.util.NoSuchElementException
	at me.champeau.gradle.japicmp.JapicmpTask.collectArchives(JapicmpTask.java:241)
	at me.champeau.gradle.japicmp.JapicmpTask.collectArchives(JapicmpTask.java:243)
	at me.champeau.gradle.japicmp.JapicmpTask.inferArchives(JapicmpTask.java:222)
	at me.champeau.gradle.japicmp.JapicmpTask.exec(JapicmpTask.java:86)
```

--- a/src/main/java/me/champeau/gradle/japicmp/JapicmpTask.java
+++ b/src/main/java/me/champeau/gradle/japicmp/JapicmpTask.java
@@ -9,6 +9,7 @@ import org.gradle.api.DefaultTask;
 import org.gradle.api.JavaVersion;
 import org.gradle.api.Project;
 import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.ResolvedArtifact;
 import org.gradle.api.artifacts.ResolvedDependency;
 import org.gradle.api.artifacts.dsl.DependencyHandler;
 import org.gradle.api.file.ConfigurableFileCollection;
@@ -238,7 +239,10 @@ public class JapicmpTask extends DefaultTask {
 
     private void collectArchives(final List<JApiCmpWorkerAction.Archive> archives, ResolvedDependency resolvedDependency) {
         String version = resolvedDependency.getModule().getId().getVersion();
-        archives.add(new JApiCmpWorkerAction.Archive(resolvedDependency.getAllModuleArtifacts().iterator().next().getFile(), version));
+        Set<ResolvedArtifact> allModuleArtifacts = resolvedDependency.getAllModuleArtifacts();
+        if (!allModuleArtifacts.isEmpty()) {
+            archives.add(new JApiCmpWorkerAction.Archive(allModuleArtifacts.iterator().next().getFile(), version));
+        }
         for (ResolvedDependency dependency : resolvedDependency.getChildren()) {
             collectArchives(archives, dependency);
         }

--- a/src/test/groovy/me/champeau/gradle/BomFunctionalTest.groovy
+++ b/src/test/groovy/me/champeau/gradle/BomFunctionalTest.groovy
@@ -1,0 +1,24 @@
+package me.champeau.gradle
+
+import org.gradle.testkit.runner.TaskOutcome
+
+class BomFunctionalTest extends BaseFunctionalTest {
+    String testProject = 'bom-project'
+
+    def "skip if module has not any artifact"() {
+        when:
+        def result = run 'japicmp'
+
+        then:
+        result.task(":japicmp").outcome == TaskOutcome.SUCCESS
+        hasTextReport('UNCHANGED CLASS: PUBLIC org.apache.commons.lang3.AnnotationUtils')
+        noHtmlReport()
+        noRichReport()
+
+        when:
+        result = run 'japicmp'
+
+        then:
+        result.task(":japicmp").outcome == TaskOutcome.UP_TO_DATE
+    }
+}

--- a/src/test/test-projects/bom-project/bom/build.gradle
+++ b/src/test/test-projects/bom-project/bom/build.gradle
@@ -1,0 +1,14 @@
+plugins {
+    id 'me.champeau.gradle.japicmp'
+    id 'java-platform'
+}
+
+repositories {
+    mavenCentral()
+}
+
+dependencies {
+    constraints {
+        api project(":simple")
+    }
+}

--- a/src/test/test-projects/bom-project/build.gradle
+++ b/src/test/test-projects/bom-project/build.gradle
@@ -1,0 +1,7 @@
+plugins {
+    id 'me.champeau.gradle.japicmp'
+}
+
+repositories {
+    mavenCentral()
+}

--- a/src/test/test-projects/bom-project/settings.gradle
+++ b/src/test/test-projects/bom-project/settings.gradle
@@ -1,0 +1,2 @@
+include ':bom'
+include ':simple'

--- a/src/test/test-projects/bom-project/simple/build.gradle
+++ b/src/test/test-projects/bom-project/simple/build.gradle
@@ -1,0 +1,7 @@
+plugins {
+    id 'me.champeau.gradle.japicmp'
+}
+
+repositories {
+    mavenCentral()
+}

